### PR TITLE
Clean up HAProxy sockets after reading

### PR DESCRIPTION
**Pull request by brianwolfe**
HAProxy closes its connection to the socket as soon as it finishes writing. This PR explicitly closes the socket from the synapse side. Without an explicit close, cleanup done on object finalization during garbage collection. This can cause synapse to temporarily accumulate a large number of socket file descriptors.

On a synthetic test with a large number of backends, this patch reduces the maximum number of open file descriptors from 940 to 52.

There are minor logging changes associated with this patch:
* added logging of error if an exception is raised when closing the file.
* if reading from the socket fails when called from https://github.com/airbnb/synapse/compare/master...brianwolfe:wolfe-socket-fd-cleanup?expand=1#diff-b2f6400f899968d5a5084ee66bf90b21R1106, there will be 2 error messages raised instead of 1, one message for the failure to read, one message for the failed command. Previously, only the failure to read error would be logged.

For old ruby versions, this cleanup can avoid this issue: https://bugs.ruby-lang.org/issues/8080 where glibc FORTIFY_SOURCE chokes on more than 1024 file descriptors despite linux supporting more.

@igor47

### CODE REVIEW

**Comment by brianwolfe** [comment]
correction: reduction was to a maximum 53 file descriptors, not 52.

File: lib/synapse/haproxy.rb [hunk 1]
```diff
@@ -1030,19 +1030,33 @@ def generate_backend_stanza(watcher, config)
...
+      nil
+    ensure
+      unless s.nil?
+        begin
+          s.close
+        rescue StandardError => e
+          log.warn "synapse: unhandled error closing stats socket: #{e.inspect}"
+        end
+      end
+      info
```
Comment (igor47) [inline lib/synapse/haproxy.rb]: if you're thinking that this will return `info` from the function since the `ensure` block is the last one that runs, that's incorrect. this does nothing, and can be omitted. example:
```
irb(main):001:0> a = begin
irb(main):002:1* 5
irb(main):003:1> ensure
irb(main):004:1* 'hi'
irb(main):005:1> end
=> 5
irb(main):006:0> a
=> 5
```

File: lib/synapse/haproxy.rb [hunk 2]
```diff
@@ -1030,19 +1030,33 @@ def generate_backend_stanza(watcher, config)
       ]
     end
 
+    def haproxy_exec(command)
+      info = nil
```
Comment (igor47) [inline lib/synapse/haproxy.rb]: i think you don't need to initialize this here; this seems like it was done just to have something for the `ensure` block to reference

File: lib/synapse/haproxy.rb [hunk 3]
```diff
@@ -1030,19 +1030,33 @@ def generate_backend_stanza(watcher, config)
...
+      s = UNIXSocket.new(@opts['socket_file_path'])
+      s.write(command)
+      info = s.read
+    rescue StandardError => e
```
Comment (igor47) [inline lib/synapse/haproxy.rb]: i'm not sure about this `rescue` clause. i think we should allow the caller to decide how to handle exceptions, even though it does make the code more verbose at the call site. i can definitely envision cases where i want to query haproxy via the stats socket but *not* want to restart haproxy if that fails

File: lib/synapse/haproxy.rb [hunk 4]
```diff
@@ -1030,19 +1030,33 @@ def generate_backend_stanza(watcher, config)
...
+      log.warn "synapse: unhandled error reading stats socket: #{e.inspect}"
+      @restart_required = true
+      nil
+    ensure
+      unless s.nil?
```
Comment (schleyfox) [inline lib/synapse/haproxy.rb]: `if s` might as well use ruby truthiness.

File: lib/synapse/haproxy.rb [hunk 5]
```diff
@@ -1089,18 +1103,10 @@ def update_backends(watchers)
...
           # actually write the command to the socket
-          begin
-            s = UNIXSocket.new(@opts['socket_file_path'])
-            s.write(command)
-            output = s.read()
-          rescue StandardError => e
-            log.warn "synapse: unknown error writing to socket"
+          output = haproxy_exec(command)
+          unless output == "\n"
+            log.warn "synapse: socket command #{command} failed: #{output}"
```
Comment (igor47) [inline lib/synapse/haproxy.rb]: although this is unrelated to the main goal of the PR, i would love to see these log lines follow the `synapse: restart required` format we use elsewhere in this function.

```ruby
log.warn "synapse: restart required because socket command #{command} failed with out #{output}"
```

File: lib/synapse/haproxy.rb [hunk 6]
```diff
@@ -1030,19 +1030,33 @@ def generate_backend_stanza(watcher, config)
...
-        s.write("show stat\n")
-        info = s.read()
-      rescue StandardError => e
-        log.warn "synapse: unhandled error reading stats socket: #{e.inspect}"
-        @restart_required = true
-        return
-      end
+      info = haproxy_exec("show stat\n")
+
+      return if info.nil?
```
Comment (igor47) [inline lib/synapse/haproxy.rb]: i was confused here, because we appear to just exit without setting `@restart_required`. i suppose the expectation is that we only get `nil` output if there was an error, but this is why i was uncertain about the broad `rescue` block in the `haproxy_exec` function.

**Comment by igor47** [comment]
okay, done with pass, let me know when you're ready for another look!

File: lib/synapse/haproxy.rb (hunk 3 above)
Comment (brianwolfe) [inline lib/synapse/haproxy.rb]: Ambivalent about it. I would lean towards having a helper method that does not trigger the restart in that case to keep it grouped where possible, but I'll pull up + reduplicate the rescue code for now.

File: lib/synapse/haproxy.rb [hunk 7]
```diff
@@ -1030,16 +1030,26 @@ def generate_backend_stanza(watcher, config)
       ]
     end
 
+    def haproxy_exec(command)
+      s = UNIXSocket.new(@opts['socket_file_path'])
+      s.write(command)
+      info = s.read
+    ensure
+      if s
+        s.close
```
Comment (igor47) [inline lib/synapse/haproxy.rb]: tighter: `s.close if s`

File: lib/synapse/haproxy.rb [hunk 8]
```diff
@@ -1030,16 +1030,26 @@ def generate_backend_stanza(watcher, config)
...
     # tries to set active backends via haproxy's stats socket
     # because we can't add backends via the socket, we might still need to restart haproxy
     def update_backends(watchers)
       # first, get a list of existing servers for various backends
       begin
-        s = UNIXSocket.new(@opts['socket_file_path'])
-        s.write("show stat\n")
-        info = s.read()
+        stat_command = "show stat\n"
+        info = haproxy_exec(stat_command)
```
Comment (igor47) [inline lib/synapse/haproxy.rb]: maybe do `info = haproxy_exec(stat_command) || ""` ? that way, if info is something funky, we'll just restart below because we don't find any backends, rather than hitting an error at `info.split`

Comment (brianwolfe) [inline lib/synapse/haproxy.rb]: shouldn't be necessary. .read should just return empty string on no backends anyway.

**Comment by igor47** [comment]
lgtm! i'll make a follow-up pr to bump the synapse version

File: lib/synapse/haproxy.rb (hunk 8 above)
Comment (brianwolfe) [inline lib/synapse/haproxy.rb]: http://ruby-doc.org/core-1.9.3/IO.html#method-i-read

> At end of file, it returns nil or "" depend on length. ios.read() and ios.read(nil) returns "". ios.read(positive-integer) returns nil.

We are not passing in a length argument, so we should get "".
